### DISCUSSION TITLE
render icon only for visible window

### DISCUSSION THREAD

**Discussion by blueray453** [davatorium/rofi#1876] [Q&A] [answered]
I have a script like bellow:

```
#!/usr/bin/env bash

shopt -s lastpipe

indexDirectory1() {
    updatedb -l 0 --database-root /media/ismail/8TBRaid0 --prunepaths /media/ismail/8TBRaid0/_Backup --prunenames ".Trash-1000 lost+found .git objects" --output $HOME/plocate_8TBRaid0.db
}

indexDirectory2() {
    updatedb -l 0 --database-root /media/ismail/SSDWorking/_Working --prunepaths /media/ismail/SSDWorking/Yandex.Disk --prunenames ".git" --output $HOME/plocate_SSDWorking_Working.db
}

indexDirectory3() {
    updatedb -l 0 --database-root /media/ismail/SSDWorking/book-collection --add-prunenames ".git" --output $HOME/plocate_SSDWorking_book_collection.db
}

export -f indexDirectory1
export -f indexDirectory2
export -f indexDirectory3

parallel -j3 ::: indexDirectory1 indexDirectory2 indexDirectory3

indexDirectoryCombine1 () {
    plocate -A --ignore-case --database $HOME/plocate_8TBRaid0.db \*
}

indexDirectoryCombine2() {
    plocate -A --ignore-case --database $HOME/plocate_SSDWorking_Working.db \*
}

indexDirectoryCombine3() {
    plocate -A --ignore-case --database $HOME/plocate_SSDWorking_book_collection.db \*
}

export -f indexDirectoryCombine1
export -f indexDirectoryCombine2
export -f indexDirectoryCombine3

parallel  --keep-order -j3 ::: indexDirectoryCombine1 indexDirectoryCombine2 indexDirectoryCombine3 | rofi -dmenu -i -theme-str 'window {fullscreen:true;}' | read select

if [ ${PIPESTATUS[1]} -eq 0 ]; then
    nemo "$select"
fi
```
I am using it as [fsearch](https://github.com/cboxdoerfer/fsearch) / [search everything](https://www.voidtools.com/) alternative. It is very fast. After first load, it takes maximum **2/3 seconds to load**.

My issue is I would like to have icons.

If I use the following code:

```
while IFS= read -r line; do
    icon="$(get_icon_from_path "$line")"
    echo -en "$line"
    echo -en "\0icon\x1f${icon}\n"
done < <(parallel -j3 ::: indexDirectoryCombine1 indexDirectoryCombine2 indexDirectoryCombine3)
```

Instead of:

    parallel  --keep-order -j3 ::: indexDirectoryCombine1 indexDirectoryCombine2 indexDirectoryCombine3

It gives me icons.

Here `get_icon_from_path` is a gjs script. It looks like:

```
#!/usr/bin/env gjs

const { Gio } = imports.gi;

let file = Gio.File.new_for_path(ARGV[0]);
let fileInfo = file.query_info('*', Gio.FileQueryInfoFlags.NONE, null);
let icons = fileInfo.get_icon().get_names();

if (typeof icons[1] === 'undefined') {
    // does not exist
    print(icons[0]);
}
else {
    print(icons[1]);
}
```
The problem is there are 4,00,000+ files in my machine. If I render icons for each calling `get_icon_from_path` it takes ages. I could not even benchmark. Cancelled benchmark after **~10 minutes**.

Is there any way I only render icon for visible window. I mean I can see ~40 rows. I would like to render icon for them only. If i use a search term. Then there will be new rows. I will only render icon for them. Just to clarify, there can be less than ~40 rows based on search term.  

Is it possible?

**Comment by DaveDavenport** [comment, marked as answer]
rofi does not render icons that are not visible, icons are lazy loaded in the background.


In git there is 'recursive browser'  this kinda does something similar and should be fast (its native in rofi), you could give that try.

**Reply by blueray453** [reply]
Thank you very much.

**Reply by blueray453** [reply]
One little confusion.

I have a file like:

```
/media/ismail/SSDWorking/book-collection/_Books/_software/architecture-design-basics
/media/ismail/SSDWorking/book-collection/_Books/business-next/seo-smm
/media/ismail/SSDWorking/book-collection/_Books/kids
/media/ismail/SSDWorking/book-collection/_Books/self-development-anxiety-self-talk
/media/ismail/SSDWorking/book-collection/_Books/self-development-anxiety-self-talk/child
```
I am calling it using:

```
#!/usr/bin/env bash

shopt -s lastpipe

cat $HOME/.dotfiles/rofi-bookmarks-list | rofi -dmenu -i -theme-str 'window {fullscreen:true;}' -show-icons | read select

if [ ${PIPESTATUS[1]} -eq 0 ]; then
    nemo "$select"
fi
```

You can see that I am using `-show-icons` but it is not showing the icons.

I searched a little and found https://github.com/davatorium/rofi/issues/840

> but the question is how to relate what icon to show with what row.

So, these are file paths and we can extract their icon information using `get_icon_from_path`. How can I tell rofi that for each row get icon using `get_icon_from_path`.

**Reply by DaveDavenport** [reply]
I don't understand. What icons? you don't seem to specify one?

**Reply by blueray453** [reply]
Sorry, for the late reply. I have been sick. Recovering but still weak. 

Let me show you what I have done so that you understand the issue. I have a script named `get_icon_from_path_stdin`. It takes a file or directory path and gives it's icon name.

```
#!/usr/bin/env gjs

'use strict';

const Gio = imports.gi.Gio;

let stdoutStream = new Gio.DataInputStream({
    base_stream: new Gio.UnixInputStream({ fd: 0 }),
    close_base_stream: true
});

while(true) {
    let [line, l] = stdoutStream.read_line_utf8(null);
    if (line == null) {break;}
    else{
        let file = Gio.File.new_for_path(line);
        let fileInfo = file.query_info('*', Gio.FileQueryInfoFlags.NONE, null);
        let icons = fileInfo.get_icon().get_names();
        if (typeof icons[1] === 'undefined') {
            print(icons[0]);
        }
        else {
            print(icons[1]);
        }
    }
}
```
I have a file which has the file paths in `$HOME/.dotfiles/rofi-bookmarks-list`, like bellow:

```
/home/ismail/.dotfiles
/home/ismail/.dotfiles/.common-dotfiles/.config/fish
/home/ismail/.dotfiles/.common-dotfiles/.local/share/applications
/home/ismail/.dotfiles/.common-dotfiles/.local/share/nemo/actions
/home/ismail/.dotfiles/.debian-dotfiles/.local/share/gnome-shell/extensions
```
I view these bookmarks with icon like:

```
#!/usr/bin/env bash

shopt -s lastpipe

coproc get_icon_from_path_stdin

exec 10>&"${COPROC[1]}"
exec 11<&"${COPROC[0]}"

while IFS= read -r line || [ -n "$line" ]; do
    echo "$line" >&10
    read -r icon <&11
    echo -en "$line"
    echo -en "\0icon\x1f${icon}\n"
done < <(cat $HOME/.dotfiles/rofi-bookmarks-list) | rofi -dmenu -i -theme-str 'window {fullscreen:true;}' | read select

if [ ${PIPESTATUS[1]} -eq 0 ]; then
    nemo "$select"
fi
```

This trick work for limited number of lines. But when I am talking about large filesystem index (400,000+ file paths), run enough times, this will most probably burn my machine.

So, what I want is to do is only run script for visible lines to get their icon. I mean to get ~40 rows icon instead of 400,000+ icons.

I did not understand your solution correctly, maybe i did not need to do all these. Please let me know what is the easiest and fastest solution to get icons for all these rows where rows contain file paths.

**Reply by DaveDavenport** [reply]
Ignoring rofi, as it looks this is not where the 'slowness' originates from:
I don't think what you want can be made fast in the method you implemented it.

~For example, you need to launch a 'gjs' instance for every file you come across, just to get an icon name.~

sorry this is wrong, I see you already have one process running.

I think this is still valid(ish) as you avoid overhead:

If you want something that is quick, it needs to be optimized to avoid launching many processes constantly and avoid doing many many filesystem accesses. Try writing the whole script in a single 'gjs' program (that walks the directories and gets the icons), that should already speed it up significantly.


If you really want it fast (and the desired asynchronous behaviour) for 400k+ files you need to write something in a language native or near native to rofi as a plugin.
As an example: If you want to get a list of f.e. all the files in your home directory, rofi (git version) has 'recursive' browser that (in the background) walks through all the files and displays these.. It can do (on a hot disk cache) 200k items in less then a second while being interactive.

**Reply by blueray453** [reply]
> If you want to get a list of f.e. all the files in your home directory, rofi (git version) has 'recursive' browser that (in the background) walks through all the files and displays these.

You were talking about https://github.com/davatorium/rofi/blob/next/source/modes/recursivebrowser.c

What is the command to list all the files in `/media/ismail/SSDWorking` with icons?

**Reply by DaveDavenport** [reply]
```bash
rofi -show recursivebrowser -recursivebrowser-directory '/tmp'
```
replace `/tmp` with the directory of choice.

**Reply by blueray453** [reply]
Thank you very much. 

I have found https://github.com/davatorium/rofi/blob/next/doc/rofi.1.markdown#recursive-browser-settings

How to set the rules:

Index:
/media/ismail/8TBRaid0
/media/ismail/SSDWorking/_Working
/media/ismail/SSDWorking/book-collection

Exclude Path:
/media/ismail/8TBRaid0/_Backup
/media/ismail/8TBRaid0/_QEMU

Exclude Folders Named:
.git
.Trash-1000
lost+found
